qcodes/tests/test_autoloadable_channels.py

```diff
@@ -0,0 +1,186 @@
+"""
+Test the auto-loadable channels and channels list. These channels are helpful
+when dealing with channel types which can be added or deleted from an
+instrument. Please note that `channel` in this context does not necessarily
+mean a physical instrument channel, but rather an instrument sub-module.
+"""
+
+from typing import List, Union
+import pytest
+import re
+
+from qcodes import Instrument
+from qcodes.instrument.channel import (
+    AutoLoadableInstrumentChannel, AutoLoadableChannelList, InstrumentChannel
+)
+
+
+class DummyBackendBase:
+    def __init__(self):
+        self._command_dict = {}
+
+    def send(self, cmd):
+        keys = self._command_dict.keys()
+        ans = None
+        key = ""
+        for key in keys:
+            ans = re.match(key, cmd)
+            if ans is not None:
+                break
+
+        if ans is None:
+            raise ValueError(f"Command {cmd} unknown")
+
+        args = ans.groups()
+        return self._command_dict[key](*args)
+
+
+class TestInstrumentBackend(DummyBackendBase):
+    def __init__(self):
+        super().__init__()
+        self._channel_catalog = ["1", "2", "4"]
+
+        self._command_dict = {
+            r":INST:CHN(\d):HLO": lambda chn: f"Hello from channel {chn}",
+            r":INST:CHN:ADD (\d)": self._channel_catalog.append,
+            r":INST:CHN:DEL (\d)": self._channel_catalog.remove,
+            r":INST:CHN:CAT": lambda: ",".join([str(i) for i in self._channel_catalog])
+        }
+
+
+class SimpleTestChannel(AutoLoadableInstrumentChannel):
+    """
+    A channel to test if we can create and delete channel instances
+    """
+
+    @classmethod
+    def _discover_from_instrument(
+            cls, parent: Instrument, **kwargs) -> List[dict]:
+        """
+        New channels need `name` and `channel` keyword arguments.
+        """
+        channels_str = parent.channel_catalog()
+        kwarg_list = [
+            {"name": f"channel{i}", "channel": int(i)}
+            for i in channels_str.split(",")
+        ]
+
+        return kwarg_list
+
+    @classmethod
+    def get_new_instance_kwargs(cls, parent: Instrument, **kwargs) -> dict:
+        """
+        Find the smallest channel number not yet occupied
+        """
+        channels_str = parent.channel_catalog()
+        existing_channels = [int(i) for i in channels_str.split(",")]
+
+        new_channel = 1
+        while new_channel in existing_channels:
+            new_channel += 1
+
+        kwargs = {
+            "name": f"channel{new_channel}",
+            "channel": new_channel
+        }
```

Comment:
in this case, technically `name` and `channel` are not kwargs :) yes, this will work fine, but it leaves a confusion. Can this be improved or at least explained in more detail(ideally, formalized in code :) )?

Reply:
Well, `name` is `channel1` for instance and `channel` is an integer `1`. Why are these not kwargs? I am not sure what additional detail is required.

Reply:
if you look at the channel's `__init__` method, you'll see that technically `name` is not a `kwarg`, but is an `arg` :) it just works because when you give a kwarg it also looks in args for it.

it would be nice if this can be explicitly mentioned in the `_get_new_instance_kwargs` docstring of the parent class.

Reply:
Ok